### PR TITLE
Fix custom path handling

### PR DESCRIPTION
**Pull request by petrkotas**
**What this PR does / why we need it**:
Improves the unintuitive container-image custom path handling. Now it fully behaves as is written in the documentation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://bugzilla.redhat.com/1660412
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubevirt/kubevirt/1891)

### CODE REVIEW

File: pkg/api/v1/deepcopy_generated.go [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // +build !ignore_autogenerated
 
 /*
-Copyright 2018 The KubeVirt Authors.
+Copyright 2019 The KubeVirt Authors.
```
Comment (stu-gott) [inline pkg/api/v1/deepcopy_generated.go]: Good catch! Rather than remove 2018, this should read:

Copyright 2018, 2019 The KubeVirt Authors.

Comment (petrkotas) [reply pkg/api/v1/deepcopy_generated.go]: The issue is, this is autogenerated. So if I change it, the travis will complain :)

**Comment by cynepco3hahue** [comment]
@petrkotas Can you please take a look on the failure
```
[Fail] ContainerDisk Starting from custom image location with disk at /custom-disk/boot.img [It] should boot normally 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:1842
```

**Comment by petrkotas** [comment]
@cynepco3hahue looking at it, double checking on local machine.

**Comment by petrkotas** [comment]
ci test please

File: tests/registry_disk_test.go [hunk 2]
```diff
@@ -160,7 +160,7 @@ var _ = Describe("ContainerDisk", func() {
 				vmi := tests.NewRandomVMIWithEphemeralDisk(tests.ContainerDiskFor(tests.ContainerDiskCirros))
 				for ind, volume := range vmi.Spec.Volumes {
 					if volume.ContainerDisk != nil {
-						vmi.Spec.Volumes[ind].ContainerDisk.Path = "/custom-disk/boot.img"
+						vmi.Spec.Volumes[ind].ContainerDisk.Path = "/custom-disk/cirros.img"
```
Comment (davidvossel) [inline tests/registry_disk_test.go]: why was this change made?

File: cmd/container-disk-v1alpha/entry-point.sh [hunk 3]
```diff
@@ -26,13 +26,16 @@ if [ -z "$COPY_PATH" ]; then
 	exit 1
 fi
 
-if [ ! -f "$IMAGE_PATH" ]; then
+if [ ! -v IMAGE_PATH ]; then
```
Comment (davidvossel) [inline cmd/container-disk-v1alpha/entry-point.sh]: why didn't the previous way work?  If the IMAGE_PATH file exists, then this logic is supposed to use the custom path, otherwise attempt to detect the file.

File: cmd/container-disk-v1alpha/entry-point.sh [hunk 4]
```diff
@@ -26,13 +26,16 @@ if [ -z "$COPY_PATH" ]; then
...
 	IMAGE_NAME=$(ls -1 disk/ | tail -n -1)
 	if [ -z "$IMAGE_NAME" ]; then
 		echo "no images found in /disk directory"
 		exit 1
 	fi
 	IMAGE_PATH=/disk/$IMAGE_NAME
+elif [ ! -e "$IMAGE_PATH" ]; then
+  echo "image not found: $IMAGE_PATH"
```
Comment (davidvossel) [inline cmd/container-disk-v1alpha/entry-point.sh]: the formatting looks inconsistent here.

File: pkg/container-disk/container-disk.go [hunk 5]
```diff
@@ -120,7 +120,7 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, po
 				// k8s minimum memory reservation is linuxMinMemory = 4194304
 				resources.Limits[kubev1.ResourceMemory] = resource.MustParse("64M")
 			}
-			containers = append(containers, kubev1.Container{
+			container := kubev1.Container{
```
Comment (davidvossel) [inline pkg/container-disk/container-disk.go]: why is this necessary

File: tests/registry_disk_test.go (hunk 2 above)
Comment (petrkotas) [reply tests/registry_disk_test.go]: Please see the dockerfile bellow. I was experimenting with negative test and it slipped into the PR. The cirros.img is the one used in dockerfile.

```dockerfile
FROM kubevirt/container-disk-v1alpha

LABEL maintainer="David Vossel <dvossel@redhat.com>"

# Add cirros
RUN curl https://download.cirros-cloud.net/0.4.0/cirros-0.4.0-x86_64-disk.img > /disk/cirros.img
RUN mkdir -p /custom-disk && cp /disk/cirros.img /custom-disk/cirros.img
```

File: pkg/container-disk/container-disk.go (hunk 5 above)
Comment (petrkotas) [reply pkg/container-disk/container-disk.go]: This is due to the need to have the IMAGE_PATH present only when the custom-path is added to VMI. Not in other case.

Comment (davidvossel) [reply pkg/container-disk/container-disk.go]: ah I see. I think I glazed over that part. thanks

**Comment by petrkotas** [comment]
@davidvossel does the changes address your remarks?

**Comment by davidvossel** [comment]
ci test please

**Comment by stu-gott** [comment]
ci test please

**Comment by petrkotas** [comment]
@stu-gott @davidvossel I see the tests failing on different test cases each time. Would you have any other wishes for this PR?

**Comment by stu-gott** [comment]
I see two of the test runs had 65 failures--at a glance they appear to be VM lifecycle related. Other than that, things do seem generally like flakey test failures.

**Comment by davidvossel** [comment]
ci test please

**Comment by davidvossel** [comment]
```

Summarizing 7 Failures:

[Fail] Windows VirtualMachineInstance [It] should succeed to start a vmi 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:1858

[Fail] Windows VirtualMachineInstance [BeforeEach] should succeed to stop a running vmi 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:343

[Fail] Windows VirtualMachineInstance with winrm connection [BeforeEach] should have correct UUID 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:1858

[Fail] Windows VirtualMachineInstance [BeforeEach] with winrm connection should have pod IP 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:343

[Fail] Windows VirtualMachineInstance with winrm connection [BeforeEach] should have the domain set properly 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:1858

[Fail] Windows VirtualMachineInstance [BeforeEach] with kubectl command should succeed to start a vmi 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:343

[Fail] Windows VirtualMachineInstance with kubectl command [It] should succeed to stop a vmi 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:1858
```

The windows functional tests appear to fail. The vms never reach the running state for some reason.

**Comment by davidvossel** [comment]
ci test please

**Comment by petrkotas** [comment]
Local run 
```
Summarizing 2 Failures:

[Fail] ContainerDisk Starting with virtio-win with virtio-win as secondary disk [It] should boot and have the virtio as sata CDROM 
/root/go/src/kubevirt.io/kubevirt/tests/registry_disk_test.go:200

[Fail] Configurations VirtualMachineInstance with CPU pinning with cpu pinning enabled [It] should start a vm with no cpu pinning after a vm with cpu pinning on same node 
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:977

Ran 207 of 248 Specs in 6140.021 seconds
FAIL! -- 205 Passed | 2 Failed | 1 Pending | 40 Skipped
```

**Comment by slintes** [comment]
ci test please

**Comment by slintes** [comment]
let's see how CI is doing today...

**Comment by slintes** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by davidvossel** [comment]
ci test please
